File: sdk-libs/token-pinocchio/src/instruction/approve.rs
//! Approve CPI for Light Token operations.

use pinocchio::{
    AccountView as AccountInfo,
    cpi::{invoke_with_slice, invoke_signed_with_slice, Signer},
    instruction::{InstructionAccount, InstructionView},
    error::ProgramError,
    address::Address,
};

use crate::constants::LIGHT_TOKEN_PROGRAM_ID;

/// Approve light-token via CPI.
///
/// # Example
///
/// ```rust,ignore
/// use light_token_pinocchio::instruction::ApproveCpi;
///
/// ApproveCpi {
///     token_account: &ctx.accounts.token_account,
///     delegate: &ctx.accounts.delegate,
///     owner: &ctx.accounts.owner,
///     system_program: &ctx.accounts.system_program,
///     amount: 100,
///     fee_payer: &ctx.accounts.fee_payer,
/// }
/// .invoke()?;
/// ```
pub struct ApproveCpi<'info> {
    pub token_account: &'info AccountInfo,
    pub delegate: &'info AccountInfo,
    pub owner: &'info AccountInfo,
    pub system_program: &'info AccountInfo,
    pub amount: u64,
    /// Fee payer for rent top-ups.
    pub fee_payer: &'info AccountInfo,
}

impl<'info> ApproveCpi<'info> {
    pub fn invoke(self) -> Result<(), ProgramError> {
        self.invoke_signed(&[])
    }

    pub fn invoke_signed(self, signers: &[Signer]) -> Result<(), ProgramError> {
        // Build instruction data: discriminator(1) + amount(8)
        let mut data = [0u8; 9];
        data[0] = 4u8; // Approve discriminator
        data[1..9].copy_from_slice(&self.amount.to_le_bytes());

        let program_id = Address::from(LIGHT_TOKEN_PROGRAM_ID);

        let account_metas = [
            InstructionAccount::writable(self.token_account.address()),
            InstructionAccount::readonly(self.delegate.address()),
            InstructionAccount::readonly_signer(self.owner.address()),
            InstructionAccount::readonly(self.system_program.address()),
            InstructionAccount::writable_signer(self.fee_payer.address()),
        ];

        let instruction = InstructionView {
            program_id: &program_id,
            accounts: &account_metas,
            data: &data,
        };

        let account_infos = [
            self.token_account,
            self.delegate,
            self.owner,
            self.system_program,
            self.fee_payer,
        ];

        if signers.is_empty() {
            invoke_with_slice(&instruction, &account_infos)
        } else {
            invoke_signed_with_slice(&instruction, &account_infos, signers)
        }
    }
}

File: js/token-sdk/src/instructions/approve.ts
/**
 * Approve and revoke delegate instructions.
 */

import type { Address } from '@solana/addresses';
import {
    AccountRole,
    type Instruction,
    type AccountMeta,
} from '@solana/instructions';

import { DISCRIMINATOR, LIGHT_TOKEN_PROGRAM_ID } from '../constants.js';
import { validatePositiveAmount } from '../utils/validation.js';
import {
    getAmountInstructionEncoder,
    getDiscriminatorOnlyEncoder,
    encodeMaxTopUp,
} from '../codecs/instructions.js';

/**
 * Parameters for approving a delegate.
 */
export interface ApproveParams {
    /** Token account to approve delegate on */
    tokenAccount: Address;
    /** Delegate to approve */
    delegate: Address;
    /** Owner of the token account - must be signer and payer */
    owner: Address;
    /** Amount to delegate */
    amount: bigint;
    /** Maximum lamports for rent top-up in units of 1,000 lamports (optional) */
    maxTopUp?: number;
}

/**
 * Creates an approve instruction (discriminator: 4).
 *
 * Approves a delegate to transfer up to the specified amount.
 *
 * Account layout:
 * 0: token account (writable)
 * 1: delegate (readonly)
 * 2: owner (signer, writable) - always the payer (APPROVE_PAYER_IDX=2 in Rust)
 *
 * Note: Unlike transfer/burn/mint-to, approve does NOT support a separate fee payer.
 * The owner is always the payer for compressible rent top-ups.
 *
 * @param params - Approve parameters
 * @returns The approve instruction
 */
export function createApproveInstruction(params: ApproveParams): Instruction {
    const { tokenAccount, delegate, owner, amount, maxTopUp } = params;

    validatePositiveAmount(amount);

    // Build accounts - owner is always WRITABLE_SIGNER (payer at index 2)
    const accounts: AccountMeta[] = [
        { address: tokenAccount, role: AccountRole.WRITABLE },
        { address: delegate, role: AccountRole.READONLY },
        { address: owner, role: AccountRole.WRITABLE_SIGNER },
    ];

    // Build instruction data: discriminator + amount [+ maxTopUp]
    const baseBytes = getAmountInstructionEncoder().encode({
        discriminator: DISCRIMINATOR.APPROVE,
        amount,
    });
    const maxTopUpBytes = encodeMaxTopUp(maxTopUp);

    const data = new Uint8Array(baseBytes.length + maxTopUpBytes.length);
    data.set(new Uint8Array(baseBytes), 0);
    if (maxTopUpBytes.length > 0) {
        data.set(maxTopUpBytes, baseBytes.length);
    }

    return {
        programAddress: LIGHT_TOKEN_PROGRAM_ID,
        accounts,
        data,
    };
}

/**
 * Parameters for revoking a delegate.
 */
export interface RevokeParams {
    /** Token account to revoke delegate from */
    tokenAccount: Address;
    /** Owner of the token account - must be signer and payer */
    owner: Address;
    /** Maximum lamports for rent top-up in units of 1,000 lamports (optional) */
    maxTopUp?: number;
}

/**
 * Creates a revoke instruction (discriminator: 5).
 *
 * Revokes the delegate authority from the token account.
 *
 * Account layout:
 * 0: token account (writable)
 * 1: owner (signer, writable) - always the payer (REVOKE_PAYER_IDX=1 in Rust)
 *
 * Note: Unlike transfer/burn/mint-to, revoke does NOT support a separate fee payer.
 * The owner is always the payer for compressible rent top-ups.
 *
 * @param params - Revoke parameters
 * @returns The revoke instruction
 */
export function createRevokeInstruction(params: RevokeParams): Instruction {
    const { tokenAccount, owner, maxTopUp } = params;

    // Build accounts - owner is always WRITABLE_SIGNER (payer at index 1)
    const accounts: AccountMeta[] = [
        { address: tokenAccount, role: AccountRole.WRITABLE },
        { address: owner, role: AccountRole.WRITABLE_SIGNER },
    ];

    // Build instruction data: discriminator [+ maxTopUp]
    const baseBytes = getDiscriminatorOnlyEncoder().encode({
        discriminator: DISCRIMINATOR.REVOKE,
    });
    const maxTopUpBytes = encodeMaxTopUp(maxTopUp);

    const data = new Uint8Array(baseBytes.length + maxTopUpBytes.length);
    data.set(new Uint8Array(baseBytes), 0);
    if (maxTopUpBytes.length > 0) {
        data.set(maxTopUpBytes, baseBytes.length);
    }

    return {
        programAddress: LIGHT_TOKEN_PROGRAM_ID,
        accounts,
        data,
    };
}
